### PR TITLE
Kusto functions support

### PR DESCRIPTION
**Pull request by barnuri**
in GetParameters the sql methods type_name, OdbcPrec, OdbcScale, OBJECT_NAME, OBJECT_SCHEMA_NAME  not supported

in GetResultElementLists the sql methods type_name, OBJECT_ID is not supported

replace all by join with relevant tables

### CODE REVIEW

**Comment by ErikEJ** [comment]
Does Kusto have functions and stored procedures??

**Comment by ErikEJ** [comment]
If not, then we should just add a check for Kusto in the scaffolder, and not attempt to scaffold procedures and functions if we are running against Kusto

**Comment by ErikEJ** [comment]
And mayb revert Kusto related changes in these classes?

File: src/Core/RevEng.Core.60/Routines/SqlServerRoutineModelFactory.cs [hunk 1]
```diff
@@ -172,22 +172,28 @@ private static Dictionary<string, List<ModuleParameter>> GetParameters(SqlConnec
             var sql = $@"
 SELECT  
     'Parameter' = p.name,  
-    'Type'   = COALESCE(type_name(p.system_type_id), type_name(p.user_type_id)),  
+    'Type'   = t.name,
```
Comment (ErikEJ) [inline src/Core/RevEng.Core.60/Routines/SqlServerRoutineModelFactory.cs:175]: This seems wrong

**Comment by barnuri** [comment]
> Does Kusto have functions and stored procedures??

no, they have function table

**Comment by ErikEJ** [comment]
Wonder if we can discover them - at any reate, we should just refarin from trying to scaffold stored procedure, so changes to SqlServerRoutineModelFactory.cs are not needed, we need to add a check in ReverseEngineerRunner instead

File: src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs
Comment (ErikEJ) [inline src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:1]: Please revert the changes, Kusto does not support stored procedures, so we should just exit early

File: src/Core/RevEng.Core.60/Routines/SqlServerRoutineModelFactory.cs
Comment (ErikEJ) [inline src/Core/RevEng.Core.60/Routines/SqlServerRoutineModelFactory.cs:1]: Please revert the changes to this file

Comment (barnuri) [reply src/Core/RevEng.Core.60/Routines/SqlServerRoutineModelFactory.cs:1]: can you please explain why ? i must this fix to make function table works
after this fix i checked it locally and be able to use function table in the db context 
i have function table with params

File: src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs [hunk 2]
```diff
@@ -44,11 +44,14 @@ protected override List<List<ModuleResultElement>> GetResultElementLists(SqlConn
             var sql = $@"
 SELECT 
     c.name,
-    COALESCE(type_name(c.system_type_id), type_name(c.user_type_id)) AS type_name,
+    t.name AS type_name,
```
Comment (ErikEJ) [inline src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:47]: This seems wrong

Comment (barnuri) [reply src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:47]: why ? its the type name i verified this change

Comment (ErikEJ) [reply src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:47]: But it can be either a user_type or a system type, but I only see this join:

`inner JOIN sys.types AS t ON c.user_type_id = t.user_type_id`

maybe you can add another:

`inner JOIN sys.types AS st ON c.system_type_id = t.system_type_id`

and then leave the COALESCE

Comment (barnuri) [reply src/Core/RevEng.Core.60/Routines/Functions/SqlServerFunctionModelFactory.cs:47]: so maybe it should be two left join because the COALESCE tell us ita can be null

**Comment by ErikEJ** [comment]
Have you managed to smoke test this with Kusto AND a SQL database with functions?

**Comment by barnuri** [comment]
> Have you managed to smoke test this with Kusto AND a SQL database with functions?

managed to run e2e on kusto, and run the specific query on sql server online emulator (doesnt have real sql server)

**Comment by ErikEJ** [comment]
@barnuri You can spin up AdventureWorks on an Azure SQL database, it has some SQL functions

**Comment by ErikEJ** [comment]
@barnuri I will try this PR branch against Adventureworks (or another database with Functions and stored procedures)

**Comment by ErikEJ** [comment]
Just tested with Adventureworks, and getting errors - I am reluctant to accept this PR, the queries should be left alone for non_Kusto usage, suggest that a variation for Kusto is used instead.

**Comment by barnuri** [comment]
> @barnuri You can spin up AdventureWorks on an Azure SQL database, it has some SQL functions

looks fine :)
image
image

**Comment by ErikEJ** [comment]
Can you find accountnumber ?

**Comment by ErikEJ** [comment]
Regardsless, there are too many changes to these two queries, that have worked well and contain a lot of knowledge

**Comment by barnuri** [comment]
> accountnumber

image

**Comment by ErikEJ** [comment]
These two queries return wildliy dfferent results against AdventureWorks:

```sql
SELECT  
    'Parameter' = p.name,  
    'Type'   = COALESCE(ts.name, tu.name),
    'Length'   = CAST(p.max_length AS INT),  
    'Precision'   = CASE 
              WHEN ts.name = 'uniqueidentifier' THEN p.precision  
              WHEN ts.name IN ('decimal', 'numeric') THEN p.precision
              WHEN ts.name IN ('varchar', 'nvarchar') THEN p.max_length
              ELSE NULL
            END, 
    'Scale'   = CAST(p.scale AS INT),  
    'Order'  = CAST(p.parameter_id AS INT),  
    p.is_output AS output,
    'TypeName' = QUOTENAME(s.name) + '.' + QUOTENAME(tu.name),
	'TypeSchema' = tu.schema_id,
	'TypeId' = p.user_type_id,
    'RoutineName' = o.name,
    'RoutineSchema' = s.name
    from sys.parameters p
    inner join sys.objects AS o on o.object_id = p.object_id
    inner JOIN sys.schemas AS s ON o.schema_id = s.schema_id
    left JOIN sys.types AS tu ON p.user_type_id = tu.user_type_id
    left JOIN sys.types AS ts ON p.system_type_id = ts.system_type_id
    ORDER BY p.object_id, p.parameter_id;
```

**Comment by barnuri** [comment]
> These two queries return wildliy dfferent results against AdventureWorks:
> 
> ```sql
> SELECT  
>     'Parameter' = p.name,  
>     'Type'   = COALESCE(ts.name, tu.name),
>     'Length'   = CAST(p.max_length AS INT),  
>     'Precision'   = CASE 
>               WHEN ts.name = 'uniqueidentifier' THEN p.precision  
>               WHEN ts.name IN ('decimal', 'numeric') THEN p.precision
>               WHEN ts.name IN ('varchar', 'nvarchar') THEN p.max_length
>               ELSE NULL
>             END, 
>     'Scale'   = CAST(p.scale AS INT),  
>     'Order'  = CAST(p.parameter_id AS INT),  
>     p.is_output AS output,
>     'TypeName' = QUOTENAME(s.name) + '.' + QUOTENAME(tu.name),
> 	'TypeSchema' = tu.schema_id,
> 	'TypeId' = p.user_type_id,
>     'RoutineName' = o.name,
>     'RoutineSchema' = s.name
>     from sys.parameters p
>     inner join sys.objects AS o on o.object_id = p.object_id
>     inner JOIN sys.schemas AS s ON o.schema_id = s.schema_id
>     left JOIN sys.types AS tu ON p.user_type_id = tu.user_type_id
>     left JOIN sys.types AS ts ON p.system_type_id = ts.system_type_id
>     ORDER BY p.object_id, p.parameter_id;
> ```
> 
> ```sql
> 	SELECT  
>     'Parameter' = p.name,  
>     'Type'   = COALESCE(type_name(p.system_type_id), type_name(p.user_type_id)),  
>     'Length'   = CAST(p.max_length AS INT),  
>     'Precision'   = CAST(case when type_name(p.system_type_id) = 'uniqueidentifier' 
>                 then p.precision  
>                 else OdbcPrec(p.system_type_id, p.max_length, p.precision) end AS INT),  
>     'Scale'   = CAST(OdbcScale(p.system_type_id, p.scale) AS INT),  
>     'Order'  = CAST(parameter_id AS INT),  
>     p.is_output AS output,
>     'TypeName' = QUOTENAME(SCHEMA_NAME(t.schema_id)) + '.' + QUOTENAME(TYPE_NAME(p.user_type_id)),
> 	'TypeSchema' = t.schema_id,
> 	'TypeId' = p.user_type_id,
>     'RoutineName' = OBJECT_NAME(p.object_id),
>     'RoutineSchema' = OBJECT_SCHEMA_NAME(p.object_id)
>     from sys.parameters p
> 	LEFT JOIN sys.table_types t ON t.user_type_id = p.user_type_id
>     ORDER BY p.object_id, p.parameter_id;
> ```

do you have suggestions ? 
from looking in the results is look like the new query return things that we didnt found before maybe its ok ?

**Comment by barnuri** [comment]
fixed the first sql

```
SELECT * INTO #temp1 FROM (
SELECT 
    c.name,
    COALESCE(ts.name, tu.name) AS type_name,
    c.column_id AS column_ordinal,
    c.is_nullable
FROM sys.columns c
inner join sys.types tu ON c.user_type_id = tu.user_type_id 
LEFT JOIN sys.types ts ON tu.system_type_id = ts.user_type_id
) as a

SELECT * INTO #temp2 FROM (
SELECT 
    c.name,
    COALESCE(type_name(c.system_type_id), type_name(c.user_type_id)) AS type_name,
    c.column_id AS column_ordinal,
    c.is_nullable
FROM sys.columns c

) as b

-- Compare the results
SELECT * FROM #temp1
EXCEPT 
SELECT * FROM #temp2;

-- Compare in the reverse direction to check if both results are identical
SELECT * FROM #temp2 
EXCEPT 
SELECT * FROM #temp1;

```

**Comment by barnuri** [comment]
@ErikEJ the first query return the same
for the second added if kusto

**Comment by barnuri** [comment]
> Leaving this until the feedback on the EF Core pr has been addressed

how it's connected ? the only left comment there is on something that i didnt changed here

**Comment by barnuri** [comment]
> Leaving this until the feedback on the EF Core pr has been addressed

update all the places from the prev PR too like i fixed in EF Core pr, found a solution with less changes the efcore team probably will like

**Comment by ErikEJ** [comment]
Cool, will have a look soon

**Comment by ErikEJ** [comment]
Have you rebased on latest master?

**Comment by barnuri** [comment]
> Have you rebased on latest master?

yeah

File: src/Core/RevEng.Core.60/PatchedSqlServerDatabaseModelFactory.cs
Comment (ErikEJ) [inline src/Core/RevEng.Core.60/PatchedSqlServerDatabaseModelFactory.cs:1]: Lets revert any changes to this file, the EF Core 6 tools does not support Kusto anyway, and I will remove EF 6 support soon anyway.

File: src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs
Comment (ErikEJ) [inline src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs:1]: Could you update the changes in this fileto align better with your EF Core PR? (Sorry, my bad for reverting)

**Review by ErikEJ — CHANGES_REQUESTED** [review]
LGTM, just a couple of changes needed

File: src/Core/RevEng.Core.60/PatchedSqlServerDatabaseModelFactory.cs
Comment (barnuri) [reply src/Core/RevEng.Core.60/PatchedSqlServerDatabaseModelFactory.cs:1]: ok, reverted

File: src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs
Comment (barnuri) [reply src/Core/RevEng.Core.80/PatchedSqlServerDatabaseModelFactory.cs:1]: done

**Review by ErikEJ — APPROVED** [review]

**Comment by ErikEJ** [comment]
Thanks @barnuri for a great contribution

**Comment by ErikEJ** [comment]
@barnuri any plans to share this via a blog post or similar??

**Comment by barnuri** [comment]
> @barnuri any plans to share this via a blog post or similar??

never done that before, if you have an example of something or write by your blog its great too
